api/logical.go

```diff
@@ -39,19 +39,60 @@ var (
 
 // Logical is used to perform logical backend operations on Vault.
 type Logical struct {
-	c *Client
+	c *LogicalContext
 }
 
-// Logical is used to return the client for logical-backend API calls.
 func (c *Client) Logical() *Logical {
-	return &Logical{c: c}
+	return &Logical{c: c.LogCtx()}
 }
 
 func (c *Logical) Read(path string) (*Secret, error) {
-	return c.ReadWithData(path, nil)
+	return c.c.Read(context.Background(), path)
 }
 
 func (c *Logical) ReadWithData(path string, data map[string][]string) (*Secret, error) {
+	return c.c.ReadWithData(context.Background(), path, data)
+}
+
+func (c *Logical) List(path string) (*Secret, error) {
+	return c.c.List(context.Background(), path)
+}
+
+func (c *Logical) Write(path string, data map[string]interface{}) (*Secret, error) {
+	return c.c.Write(context.Background(), path, data)
+}
+
+func (c *Logical) WriteBytes(path string, data []byte) (*Secret, error) {
+	return c.c.WriteBytes(context.Background(), path, data)
+}
+
+func (c *Logical) Delete(path string) (*Secret, error) {
+	return c.c.Delete(context.Background(), path)
+}
+
+func (c *Logical) DeleteWithData(path string, data map[string][]string) (*Secret, error) {
+	return c.c.DeleteWithData(context.Background(), path, data)
+}
+
+func (c *Logical) Unwrap(wrappingToken string) (*Secret, error) {
+	return c.c.Unwrap(context.Background(), wrappingToken)
+}
+
+// LogicalContext is used to perform logical backend operations on Vault with a
+// context argument.
+type LogicalContext struct {
+	c *Client
+}
+
+func (c *Client) LogCtx() *LogicalContext {
```

Comment:
I wonder if this method's name might confuse people, since to me "LogCtx" sounds like a context related to logging.

Reply:
Yeah, I was pretty iffy about the name.  I was originally going with LogicalContext() but that seemed unwieldy, since people often write stuff like

```
resp, err := c.Logical().Write("/some/rather/long/path", map[string]interface{}{
 ...
```

and between the longer name and the new ctx argument it'll just be worse.  Still, probably better to have long lines than misleading names.